### PR TITLE
[ORCA-89] Add design notes about the proof of concept to the README

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
I reviewed the Nextflow workflow and Python script and took note of important design decisions that I made when I was putting together the proof of concept. If anything is unclear in my notes or if you scan the Nextflow workflow and Python script, let me know in the comments and I can clarify!

### CODE REVIEW

**Comment by thomasyu888** [comment]
Will review this in more detail later today or tomorrow.

File: README.md [hunk 1]
```diff
@@ -379,6 +379,34 @@ Succeeded   : 50
...
+
+- The pipeline currently uses container images on Docker Hub, which might be affected by rate limits.
+- Currently, the Nextflow workflow is set up to run the external command(s) in one process, which means that we are limited to a single container. While we should be able to run multiple commands, if they require different programs, we could either build a composite/mulled containers, refactor the Nextflow workflow to allow for an arbitrary number of steps, or run a WMS (_e.g._ cwltool) within a Nextflow process. Depending on how common this need is, we can decide on which solution is most appropriate. For example, if it's a rare occurrence, then composite/mulled containers would be a reasonable approach.
+- Special care is taken to make sure that the file names are unique in order to avoid name clashes during the two aggregation processes at the end of the workflow. Currently, this involves naming files based on the Synapse ID and the QC test name, but this doesn't generalize well. For example, what would take the place of the Synapse ID in the case of a HTTPS file? Perhaps the unique identifiers can be the index of the data file in the input CSV file (_i.e._ the row number).
```
Comment (BWMac) [inline README.md:401]: I was thinking about this when working on `nf-validate` and was thinking about concatenating the run-name and the index from the csv file to create a unique identifier for the files.

**Review by BWMac — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -379,6 +379,34 @@ Succeeded   : 50
 
 </details>
 
+## Design
+
+### High-level Decisions
+
+- **Nextflow Workflow**
+  - Many QC tests will require the output from one or more commands using external (_i.e._ non-Python) programs. While we can run external commands in Python (_e.g. using the `subprocess` module), we want to avoid re-inventing a workflow management system (WMS). Instead, we should leverage an existing WMS. For the purposes of this POC, we opted for Nextflow since we have the compute infrastructure set up. That said, a future iteration could use the Common Workflow Language (whose reference implementation is in Python) and/or the Task Execution Service (TES).
```
Comment (thomasyu888) [inline README.md:387]: That's pretty interesting to leverage CWL with the combination of leveraging TES.

File: README.md [hunk 3]
```diff
@@ -379,6 +379,34 @@ Succeeded   : 50
...
+  - Given that Nextflow supports arbitrary Groovy/Java code, we could technically accomplish all of the work in Nextflow alone. However, it makes a lot more sense to use a programming language that our team (_i.e._ DPE) is familiar with and that people outside of our team can readily contribute to. Python fits the bill. Hence, we are currently building a Python-Nextflow hybrid tool. 
+  - In Nextflow, the `container` directive and the `script` block can leverage values provided as inputs (_e.g._ `container "${some_input}"`). This functionality allows us to keep the Nextflow workflow very general. Otherwise, we would have to hard-code the containers and commands in the Nextflow workflow and somehow keep everything in sync with the Python package, which takes care of interpreting the commands.
```
Comment (thomasyu888) [inline README.md:389]: I like this about Nextflow. One concern I have about this is not all Docker images are supported by Tower, @BWMac also ran into an issue like [this package](https://github.com/Sage-Bionetworks/agora-data-tools/blob/40ae5870c4c6965a4931bbb6961de76ca9f63e23/Dockerfile#L4) needing to be installed.

This is tangential to this PR, but I imagine we can potentially do this for the ITCR docker execution recipe.

File: README.md [hunk 4]
```diff
@@ -379,6 +379,34 @@ Succeeded   : 50
...
+  - The Nextflow workflow makes use of a short `parseJson()` Groovy function to parse the contents of a JSON file into a Nextflow channel. This helps bridge the gap between Python-based processes and Nextflow since Nextflow cannot easily handle multiple outputs that are not files (_e.g._ stdout), especially when the output is intended to be structured.
+- **Python Package**
+  - I use the term "QC target" to refer to any combination of files and folders that represent a type of data and can be validated as a group. As a starting point, we are focusing on single files, so "QC target" and "QC file" can be used interchangeably. However, we will quickly run into multi-file QC targets, such as FASTQ file pairs and BAM/CRAM files with their indices.
```
Comment (thomasyu888) [inline README.md:392]: An example of a "QC target" would be nice.  Just so I'm understanding, would an example of a "QC target" be "RNASeq FASTQ"?

File: README.md [hunk 5]
```diff
@@ -379,6 +379,34 @@ Succeeded   : 50
...
+  - Almost every class has a `to_dict()` method to generate a dictionary representation of an instance of that class. This is used primarily to serialize the objects to JSON between steps in the Nextflow workflow. Ideally, the Python package can deserialize these JSON files into the appropriate set of objects.
+  - In the current version, I distinguish between internal and external tests. This is used to separate tests into two branches in the Nextflow DAG.
+  - The QC suite is intended to represent the set of tests that are relevant to each file type. This approach might need to be re-engineered a bit to accommodate multi-file QC targets though. I'm also using inheritance to accumulate additional QC tests. For example, the OME-TIFF QC suite has OME-specific tests in addition to those inherited from the TIFF QC suite, which in turn inherits from the generic file QC suite (with basic tests like MD5 checksum validation).
+
+### Caveats
+
+- The pipeline currently uses container images on Docker Hub, which might be affected by rate limits.
```
Comment (thomasyu888) [inline README.md:399]: We can have a discussion as a group about Docker Hub and its future. I wrote a [document](https://docs.google.com/document/d/1GGXITF2k75VWLejpUcQ_7uSsqyN2l0mS3DPZQ2VlXu4/edit) about potentially shifting to GHCR, but there are unknowns about GitHub's application usage costs.

File: README.md [hunk 6]
```diff
@@ -379,6 +379,34 @@ Succeeded   : 50
...
+- Special care is taken to make sure that the file names are unique in order to avoid name clashes during the two aggregation processes at the end of the workflow. Currently, this involves naming files based on the Synapse ID and the QC test name, but this doesn't generalize well. For example, what would take the place of the Synapse ID in the case of a HTTPS file? Perhaps the unique identifiers can be the index of the data file in the input CSV file (_i.e._ the row number).
+
+### Future Directions
+
+- For now, the workflow is focusing on QC status (_i.e._ `pass`, `fail`, and `skip`). In the future, we can look into generating QC metrics and even build those into the relevant data models so the metrics can be captured as metadata. This may involve leveraging MultiQC.
+- We cannot leverage jobs cached by Nextflow because we cannot rely on intermediate files being stored indefinitely. Hence, we need to build in a separate method for skipping files that have already been QC'ed.
```
Comment (thomasyu888) [inline README.md:406]: I have some ideas about this here, but I'd like to hear your thoughts.  We can have a separate ticket to track this in the future.

**Review by thomasyu888 — APPROVED** [review]
Looks great!  Very thorough per usual.  I left some comments and questions, but going to approve!

File: README.md (hunk 4 above)
Comment (BrunoGrandePhD) [reply README.md:392]: I added this:

> For example, RNA-seq FASTQ file pairs and BAM/CRAM files with their indices would be represented by a `QcFilePair`, whereas a more complex group of files would be represented by a `QcFileGroup`, which keeps tracks of relative paths.

File: README.md (hunk 6 above)
Comment (BrunoGrandePhD) [reply README.md:406]: Sounds good!

**Pull request merged**